### PR TITLE
latency histogram option for arangoimport

### PR DESCRIPTION
**Pull request by matthewvon**
A new option "--latency true" will add a histogram output of latency data to arangoimport.  All timing values are in milliseconds.

### CODE REVIEW

**Comment by jsteemann** [comment]
@matthewvon : it would be nice if you could eventually start adopting our coding guidelines.

**Comment by matthewvon** [comment]
@jsteemann You have to be more specific than that.  I rewrote this with class basic::Thread.  Thought I camel cased everything.  What is giving you grief?

File: arangosh/Import/ImportHelper.h [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include <atomic>
 
 #include "AutoTuneThread.h"
+#include "quick_hist.h"
```
Comment (jsteemann) [inline arangosh/Import/ImportHelper.h]: File name convention here differs from all other file names used in ArangoDB and that are our own code.

File: arangosh/Import/quick_hist.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+#include "Basics/ConditionVariable.h"
+#include "Basics/Thread.h"
+#include "Logger/Logger.h"
+
+namespace arangodb {
+namespace import {
+//
+// quickly written histogram class for debugging
+//
+  class QuickHistogram : public arangodb::Thread {
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: indentation is somewhat counter-intuitive here?

File: arangosh/Import/quick_hist.h [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  QuickHistogram(QuickHistogram const&) = delete;
+  QuickHistogram& operator=(QuickHistogram const&) = delete;
+
+ public:
+  QuickHistogram()
+    : Thread("QuickHistogram"), _threadRunning(false)
+  {};
+
+  ~QuickHistogram()
+  {
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: brackets for this function differ from brackets for other functions. why?

File: arangosh/Import/quick_hist.h [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+//
+  class QuickHistogram : public arangodb::Thread {
+ private:
+  QuickHistogram(QuickHistogram const&) = delete;
+  QuickHistogram& operator=(QuickHistogram const&) = delete;
+
+ public:
+  QuickHistogram()
+    : Thread("QuickHistogram"), _threadRunning(false)
+  {};
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: superfluous semicolon

File: arangosh/Import/quick_hist.h [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+  ~QuickHistogram()
+  {
+    shutdown();
+  }
+
+  void beginShutdown() override {
+    _threadRunning=false;
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: `_threadRunning = false`

File: arangosh/Import/quick_hist.h [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+      _writingLatencies->push_back(latency);
+    }
+  }
+
+protected:
+  void run() override
+  {
+    _interval_start=std::chrono::steady_clock::now();
+    _measuring_start = _interval_start;
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: inconsistent use of whitespace. Line 81 uses whitespace for the assignment, and line 80 does not.
According to our coding conventions the whitespace should be used everywhere and consistently.

File: arangosh/Import/quick_hist.h [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    if (_threadRunning.load()) {
+      CONDITION_LOCKER(guard, _condvar);
+
+      _writingLatencies->push_back(latency);
+    }
+  }
+
+protected:
+  void run() override
+  {
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: brackets now in new line... our coding style puts the beginning of a block at the end of the function signature header.

File: arangosh/Import/quick_hist.h [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+      << R"("elapsed","window","n","min","mean","median","95th","99th","99.9th","max","unused1","clock")";
+
+    _writingLatencies = &_vectors[0];
+    _readingLatencies = &_vectors[1];
+    _threadRunning.store(true);
+
+    while(_threadRunning.load()) {
+      CONDITION_LOCKER(guard, _condvar);
+
+      std::chrono::seconds ten_sec(10);
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: variable naming with underscore in the middle of the name is somewhat out of our coding conventions.

File: arangosh/Import/quick_hist.h [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+      _condvar.wait(ten_sec);
+
+      LatencyVec_t * temp;
+      temp = _writingLatencies;
+      _writingLatencies = _readingLatencies;
+      _readingLatencies = temp;
+
+      print_interval(!_threadRunning.load());
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: The same for function names.

File: arangosh/Import/quick_hist.h [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    }
+  } // run
+
+
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief
+  //////////////////////////////////////////////////////////////////////////////
+  std::chrono::steady_clock::time_point _measuring_start;
+  std::chrono::steady_clock::time_point _interval_start;
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: and instance variables

File: arangosh/Import/quick_hist.h [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+          } else {
+            median = _readingLatencies->at(half);
+          }
+        } else {
+          sum = zero_micros;
+          mean = zero_micros;
+          median = zero_micros;
+        } // else
+
+        sort(_readingLatencies->begin(), _readingLatencies->end());
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: std::sort?

File: arangosh/Import/quick_hist.h [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        _readingLatencies->clear();
+        _interval_start=interval_end;
+      }
+    }
+  }
+
+  //
+  // calculation taken from http://www.dummies.com/education/math/statistics/how-to-calculate-percentiles-in-statistics
+  //  (zero and one size vector calculations not included in that link)
+  std::chrono::microseconds CalcPercentile(std::vector<std::chrono::microseconds> &SortedLatencies, int Percentile) {
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: According to our coding conventions, we put both the `&` and the `*` on the left-hand side.

Comment (jsteemann) [reply arangosh/Import/quick_hist.h]: And we wouldn't use function parameter names starting with an upper case letter.

File: arangosh/Import/quick_hist.h [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+        ret_val = SortedLatencies[0] / 2;
+      } else {
+        ret_val=std::chrono::microseconds(0);
+      } // else
+    } // else
+
+    return ret_val;
+  } // CalcPercentile
+
+ private:
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: ?

File: arangosh/Import/quick_hist.h [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+};
+
+
+class QuickHistogramTimer {
+public:
+  QuickHistogramTimer(QuickHistogram & histo)
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: explicit?

File: arangosh/Import/quick_hist.h [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    : _histogram(histo) {
+    _interval_start=std::chrono::steady_clock::now();
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: use initializer list?

File: arangosh/Import/quick_hist.h [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+  }
+
+  ~QuickHistogramTimer() {
+    std::chrono::microseconds latency;
+
+    latency = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - _interval_start);
+    _histogram.post_latency(latency);
```
Comment (jsteemann) [inline arangosh/Import/quick_hist.h]: may throw (calls push_back) and is not safe to call in a dtor.

**Comment by jsteemann** [comment]
@matthewvon : please refer to my comments

**Comment by matthewvon** [comment]
@jsteemann no sweat

**Comment by jsteemann** [comment]
@matthewvon : vale

File: arangosh/Import/quick_hist.h (hunk 16 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: you cool with a try around the call and log on catch?

File: arangosh/Import/quick_hist.h (hunk 2 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: emacs tried to help.  fixed.

File: arangosh/Import/quick_hist.h (hunk 3 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 4 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 5 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 6 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 7 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 10 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 12 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: i think fixed

File: arangosh/Import/quick_hist.h (hunk 13 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 9 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

**Comment by matthewvon** [comment]
Requested changes pushed.

File: arangosh/Import/quick_hist.h (hunk 15 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed

File: arangosh/Import/quick_hist.h (hunk 8 above)
Comment (matthewvon) [reply arangosh/Import/quick_hist.h]: fixed everywhere

**Pull request merged**